File: protocol/handshake.py
"""
Eirn-KCP Handshake: MSG0' Construction and Processing.

Implements the complete handshake protocol supporting both
KcpMode.LITE (hash-based KCP) and KcpMode.STRICT (Lattice ZKPoK).

SENDER (Alice):
  1. Generate ephemeral keypair (ek_A, esk_A)
  2. Compute standard KEM encapsulations: ct1, ct2, ct3
  3. Compute NAXOS-KEM encapsulation: ct4
  4. Compute ZK proof (Lite or Strict based on kcp_mode)
  5. Form MSG0' = (pk_A, ek_A, ct1, ct2, ct3, ct4, kcp_mode, zk_proof, [lattice_pk])
  6. Derive session key

RECEIVER (Bob):
  1. Receive MSG0'
  2. Verify zk_proof (abort if invalid)  ← FAIL-FAST
  3. Decapsulate ct1, ct2, ct3, ct4
  4. Derive session key (must match Alice's)
"""

from dataclasses import dataclass
from typing import Tuple, Optional, Union

from eirn_kem.kem import (
    KeyGen, Encaps, Decaps,
    EirnPublicKey, EirnSecretKey, EirnCiphertext
)
from eirn_kem.utils import sha3_256, concat, random_bytes
from eirn_kem.params import DEFAULT_PARAMS

from naxos_kem.naxos import naxos_encaps, naxos_decaps

from zk.sigma import prove as kcp_lite_prove, verify as kcp_lite_verify, ZKProof
from lattice_zk import (
    KcpMode, derive_lattice_keys, lattice_prove, lattice_verify,
    LatticeZKProof, LatticePublicKey
)

from .prekey import PrekeyBundle
from .session import derive_session_key


class AuthenticationError(Exception):
    """Raised when ZK proof verification fails."""
    pass


@dataclass
class MSG0Prime:
    """
    Extended initial handshake message with ZK proof.
    
    Fields:
      pk_A: Alice's identity public key
      ek_A: Alice's ephemeral public key
      ct1:  Encaps(pk_B)     — identity binding
      ct2:  Encaps(spk_B)    — prekey forward secrecy
      ct3:  Encaps(opk_B)    — one-time forward secrecy
      ct4:  NAXOS.Encaps(spk_B, sk_A, esk_A, ctx) — auth
      kcp_mode: Protocol mode (LITE or STRICT)
      zk_proof: The proof (ZKProof for LITE, LatticeZKProof for STRICT)
      lattice_pk: Alice's lattice public key (STRICT mode only)
    """
    pk_A: EirnPublicKey
    ek_A: EirnPublicKey
    ct1: EirnCiphertext
    ct2: EirnCiphertext
    ct3: EirnCiphertext
    ct4: EirnCiphertext
    kcp_mode: KcpMode
    zk_proof: Union[ZKProof, LatticeZKProof]
    lattice_pk: Optional[LatticePublicKey] = None
    
    def total_size(self) -> int:
        """Compute total message size in bytes."""
        base = (
            len(bytes(self.pk_A)) +
            len(bytes(self.ek_A)) +
            len(bytes(self.ct1)) +
            len(bytes(self.ct2)) +
            len(bytes(self.ct3)) +
            len(bytes(self.ct4)) +
            1 # mode
        )
        
        proof_size = (
            self.zk_proof.proof_size() 
            if isinstance(self.zk_proof, LatticeZKProof) 
            else 96
        )
        
        lat_pk_size = len(self.lattice_pk.to_bytes()) if self.lattice_pk else 0
        
        return base + proof_size + lat_pk_size


def build_context(
    pk_A: EirnPublicKey,
    pk_B: EirnPublicKey,
    ek_A: EirnPublicKey
) -> bytes:
    """
    Build the session context for ZK proof and NAXOS-KEM.
    """
    return concat(bytes(pk_A), bytes(pk_B), bytes(ek_A))


def sender_handshake(
    sk_A: EirnSecretKey,
    pk_A: EirnPublicKey,
    bundle_B: PrekeyBundle,
    kcp_mode: KcpMode = KcpMode.LITE
) -> Tuple[MSG0Prime, bytes]:
    """
    Alice's side of the Eirn-KCP handshake.
    Produces MSG0' and derives the session key.
    """
    pub_bundle = bundle_B.get_public_bundle()
    pk_B = pub_bundle["pk_B"]
    spk_B = pub_bundle["spk_B"]
    opk_B = pub_bundle["opk_list"][0]  # consume first OPK
    
    # Step A1: Generate ephemeral keypair
    ek_A, esk_A = KeyGen()
    
    # Build session context
    ctx = build_context(pk_A, pk_B, ek_A)
    
    # Step A2: Standard KEM encapsulations
    ct1, K1 = Encaps(pk_B)
    ct2, K2 = Encaps(spk_B)
    ct3, K3 = Encaps(opk_B)
    
    # Step A3: NAXOS-KEM
    ct4, K4 = naxos_encaps(pk_R=spk_B, sk_S=sk_A, esk_S=esk_A, context=ctx)
    
    # Step A4: ZK proof generation
    lattice_pk = None
    if kcp_mode == KcpMode.STRICT:
        lattice_pk, lattice_sk = derive_lattice_keys(bytes(sk_A))
        zk_proof = lattice_prove(lattice_sk, ctx)
        if zk_proof is None:
            raise RuntimeError("Lattice proof generation failed")
    else:
        zk_proof = kcp_lite_prove(sk_A=bytes(sk_A), pk_A=bytes(pk_A), ctx=ctx)
        
    # Step A5: Session key
    ss = derive_session_key(
        shared_secrets=[K1, K2, K3, K4],
        pk_A=bytes(pk_A),
        pk_B=bytes(pk_B),
        ek_A=bytes(ek_A),
        ciphertexts=[bytes(ct1), bytes(ct2), bytes(ct3), bytes(ct4)]
    )
    
    msg0 = MSG0Prime(
        pk_A=pk_A,
        ek_A=ek_A,
        ct1=ct1,
        ct2=ct2,
        ct3=ct3,
        ct4=ct4,
        kcp_mode=kcp_mode,
        zk_proof=zk_proof,
        lattice_pk=lattice_pk
    )
    
    return msg0, ss


def receiver_handshake(
    bundle_B: PrekeyBundle,
    msg0: MSG0Prime
) -> bytes:
    """
    Bob's side of the Eirn-KCP handshake.
    VERIFIES proof BEFORE decapsulation (fail-fast).
    """
    pk_A = msg0.pk_A
    ek_A = msg0.ek_A
    
    pk_B = bundle_B.pk_B
    sk_B = bundle_B.sk_B
    ssk_B = bundle_B.ssk_B
    osk_B = bundle_B.consume_opk()[1]
    
    ctx = build_context(pk_A, pk_B, ek_A)
    
    # VERIFY ZK proof (fail-fast)
    if msg0.kcp_mode == KcpMode.STRICT:
        if msg0.lattice_pk is None:
            raise AuthenticationError("STRICT mode requires lattice_pk")
        zk_valid = lattice_verify(msg0.lattice_pk, ctx, msg0.zk_proof)
    else:
        zk_valid = kcp_lite_verify(bytes(pk_A), ctx, msg0.zk_proof)
        
    if not zk_valid:
        raise AuthenticationError("ZK proof verification failed")
    
    # Decapsulations
    K1 = Decaps(sk_B, msg0.ct1)
    K2 = Decaps(ssk_B, msg0.ct2)
    K3 = Decaps(osk_B, msg0.ct3)
    K4 = naxos_decaps(sk_R=ssk_B, ct=msg0.ct4, pk_S=pk_A, context=ctx)
    
    # Session key derivation
    ss = derive_session_key(
        shared_secrets=[K1, K2, K3, K4],
        pk_A=bytes(pk_A),
        pk_B=bytes(pk_B),
        ek_A=bytes(ek_A),
        ciphertexts=[bytes(msg0.ct1), bytes(msg0.ct2), bytes(msg0.ct3), bytes(msg0.ct4)]
    )
    
    return ss

File: src/handshake.rs
//! Sender and receiver handshake orchestration for Eirn-KCP.
//!
//! This module constructs and consumes the initial `MSG0'` message. It is
//! designed to resist replay and key-substitution attacks when prekey bundles
//! are authenticated and one-time prekeys are consumed exactly once. It does not
//! protect against maliciously supplied receiver bundles or side channels in the
//! surrounding application.

use crate::{
    error::{EirnError, Result},
    kem::{decaps, encaps, keygen, Ciphertext, PublicKey},
    naxos::{naxos_decaps, naxos_encaps},
    prekey::{PrekeyBundle, PublicPrekeyBundle},
    session::derive_session_key,
    zk::{kcp_lite_prove_for_key, kcp_lite_verify, KcpLiteProof, KcpMode},
};

/// Initial sender message for the asynchronous Eirn-KCP handshake.
///
/// The message carries the sender identity key, sender ephemeral key, four
/// ciphertexts with distinct binding roles, and a KCP-Lite proof tying the
/// sender identity to the session context. Verifying the message authenticates
/// the sender only if the sender public key is expected by the caller. Callers
/// must reject unauthenticated receiver bundles and must not reuse consumed
/// one-time prekeys.
#[derive(Clone, Debug)]
pub struct Msg0Prime {
    pub pk_a: PublicKey,
    pub ek_a: PublicKey,
    pub ct1: Ciphertext,
    pub ct2: Ciphertext,
    pub ct3: Ciphertext,
    pub ct4: Ciphertext,
    pub kcp_mode: KcpMode,
    pub kcp_lite_proof: KcpLiteProof,
}

impl Msg0Prime {
    /// Fixed encoded `MSG0'` size in bytes.
    pub const SIZE: usize = 1 + PublicKey::SIZE * 2 + Ciphertext::SIZE * 4 + KcpLiteProof::SIZE;

    /// Returns the fixed encoded size for this message.
    ///
    /// # Security
    ///
    /// The size is structural metadata only. Callers must still verify the
    /// KCP-Lite proof and derive the session key before accepting a message.
    pub fn total_size(&self) -> usize {
        Self::SIZE
    }

    /// Serializes the message to its fixed-size wire encoding.
    ///
    /// # Security
    ///
    /// Serialization does not authenticate the message. Receivers must run
    /// [`receiver_handshake`] or equivalent proof verification before using
    /// derived keys.
    pub fn to_bytes(&self) -> [u8; Self::SIZE] {
        let mut out = [0u8; Self::SIZE];
        out[0] = self.kcp_mode as u8;

        let mut offset = 1;
        write_field(&mut out, &mut offset, self.pk_a.as_bytes());
        write_field(&mut out, &mut offset, self.ek_a.as_bytes());
        write_field(&mut out, &mut offset, self.ct1.as_bytes());
        write_field(&mut out, &mut offset, self.ct2.as_bytes());
        write_field(&mut out, &mut offset, self.ct3.as_bytes());
        write_field(&mut out, &mut offset, self.ct4.as_bytes());
        write_field(&mut out, &mut offset, &self.kcp_lite_proof.to_bytes());
        out
    }

    /// Parses a default-profile message from its fixed-size wire encoding.
    ///
    /// # Errors
    ///
    /// Returns [`EirnError::InvalidMessageLength`] if `data` has the wrong
    /// length, [`EirnError::StrictModeUnavailable`] if the mode byte requests
    /// strict KCP, or a key, ciphertext, or proof length error if a field fails
    /// structural parsing.
    ///
    /// # Untrusted Input
    ///
    /// This function accepts data from untrusted sources. All structural checks
    /// are performed before any arithmetic. Malformed input is rejected with an
    /// error rather than panicking.
    ///
    /// # Security
    ///
    /// Parsing does not verify the sender proof or consume a one-time prekey.
    /// Callers must pass the result to [`receiver_handshake`] before accepting
    /// the transcript.
    pub fn from_bytes(data: &[u8]) -> Result<Self> {
        Self::from_bytes_with_params(data, crate::params::PARAMS_512)
    }

    /// Parses a message from its fixed-size wire encoding under `params`.
    ///
    /// # Errors
    ///
    /// Returns [`EirnError::InvalidMessageLength`] if `data` has the wrong
    /// length, [`EirnError::StrictModeUnavailable`] if the mode byte requests
    /// strict KCP, or a key, ciphertext, or proof length error if a field fails
    /// structural parsing.
    ///
    /// # Untrusted Input
    ///
    /// This function accepts data from untrusted sources. All structural checks
    /// are performed before any arithmetic. Malformed input is rejected with an
    /// error rather than panicking.
    ///
    /// # Security
    ///
    /// The caller must supply the same parameter profile expected for the peer
    /// bundle. Successful parsing is not proof verification.
    pub fn from_bytes_with_params(data: &[u8], params: crate::params::EirnParams) -> Result<Self> {
        if data.len() != Self::SIZE {
            return Err(EirnError::InvalidMessageLength {
                expected: Self::SIZE,
                actual: data.len(),
            });
        }

        if data[0] != KcpMode::Lite as u8 {
            return Err(EirnError::StrictModeUnavailable);
        }

        let mut offset = 1;
        let pk_a =
            PublicKey::from_bytes(read_field::<{ PublicKey::SIZE }>(data, &mut offset), params)?;
        let ek_a =
            PublicKey::from_bytes(read_field::<{ PublicKey::SIZE }>(data, &mut offset), params)?;
        let ct1 = Ciphertext::from_bytes(
            read_field::<{ Ciphertext::SIZE }>(data, &mut offset),
            params,
        )?;
        let ct2 = Ciphertext::from_bytes(
            read_field::<{ Ciphertext::SIZE }>(data, &mut offset),
            params,
        )?;
        let ct3 = Ciphertext::from_bytes(
            read_field::<{ Ciphertext::SIZE }>(data, &mut offset),
            params,
        )?;
        let ct4 = Ciphertext::from_bytes(
            read_field::<{ Ciphertext::SIZE }>(data, &mut offset),
            params,
        )?;
        let kcp_lite_proof =
            KcpLiteProof::from_bytes(read_field::<{ KcpLiteProof::SIZE }>(data, &mut offset))?;

        Ok(Self {
            pk_a,
            ek_a,
            ct1,
            ct2,
            ct3,
            ct4,
            kcp_mode: KcpMode::Lite,
            kcp_lite_proof,
        })
    }
}

/// Builds the KCP-Lite session context from sender, receiver, and ephemeral keys.
///
/// # Security
///
/// The context is only meaningful when `pk_b` is the authenticated receiver
/// identity key and `ek_a` is fresh for the session. Reusing `ek_a` weakens
/// replay separation across handshakes.
pub fn build_context(pk_a: &PublicKey, pk_b: &PublicKey, ek_a: &PublicKey) -> Vec<u8> {
    let mut ctx = Vec::with_capacity(96);
    ctx.extend_from_slice(pk_a.as_bytes());
    ctx.extend_from_slice(pk_b.as_bytes());
    ctx.extend_from_slice(ek_a.as_bytes());
    ctx
}

/// Creates the sender's initial message and transcript-bound session key.
///
/// # Errors
///
/// Returns [`EirnError::KeyMismatch`] if `sk_a` does not own `pk_a`.
/// Returns [`EirnError::NoOneTimePrekeys`] if `bundle_b` has no public one-time
/// prekey. Returns [`EirnError::KeyMismatch`] from KCP-Lite proof generation if
/// the key relationship fails during proof construction.
///
/// # Security
///
/// `bundle_b` must be authenticated as the receiver's current public prekey
/// bundle. The function consumes no receiver state because it only sees the
/// public half; receivers must enforce one-time prekey consumption when they
/// process the message.
pub fn sender_handshake(
    sk_a: &crate::kem::SecretKey,
    pk_a: &PublicKey,
    bundle_b: &PublicPrekeyBundle,
) -> Result<(Msg0Prime, [u8; 32])> {
    if !sk_a.matches_public_key(pk_a) {
        return Err(EirnError::KeyMismatch);
    }

    let (ek_a, esk_a) = keygen();
    let opk_b = bundle_b
        .one_time_prekeys
        .first()
        .ok_or(EirnError::NoOneTimePrekeys)?
        .clone();
    let ctx = build_context(pk_a, &bundle_b.identity_pk, &ek_a);

    // ct1 binds the session to the receiver identity key.
    let (ct1, k1) = encaps(&bundle_b.identity_pk);
    // ct2 binds the session to the receiver signed prekey.
    let (ct2, k2) = encaps(&bundle_b.signed_prekey_pk);
    // ct3 consumes a one-time prekey for replay separation.
    let (ct3, k3) = encaps(&opk_b);
    // ct4 adds NAXOS-style sender identity and ephemeral entropy.
    let (ct4, k4) = naxos_encaps(&bundle_b.signed_prekey_pk, sk_a, &esk_a, &ctx);

    let proof = kcp_lite_prove_for_key(sk_a, pk_a, &ctx)?;
    let secrets = [k1, k2, k3, k4];
    let ciphertexts = [
        ct1.as_bytes().as_slice(),
        ct2.as_bytes().as_slice(),
        ct3.as_bytes().as_slice(),
        ct4.as_bytes().as_slice(),
    ];
    let ss = derive_session_key(
        &secrets,
        pk_a.as_bytes(),
        bundle_b.identity_pk.as_bytes(),
        ek_a.as_bytes(),
        &ciphertexts,
        None,
    );

    Ok((
        Msg0Prime {
            pk_a: pk_a.clone(),
            ek_a,
            ct1,
            ct2,
            ct3,
            ct4,
            kcp_mode: KcpMode::Lite,
            kcp_lite_proof: proof,
        },
        ss,
    ))
}

fn write_field<const N: usize>(out: &mut [u8], offset: &mut usize, field: &[u8; N]) {
    out[*offset..*offset + N].copy_from_slice(field);
    *offset += N;
}

fn read_field<'a, const N: usize>(data: &'a [u8], offset: &mut usize) -> &'a [u8] {
    let field = &data[*offset..*offset + N];
    *offset += N;
    field
}

/// Verifies `msg0`, consumes one one-time prekey, and derives the receiver key.
///
/// # Errors
///
/// Returns [`EirnError::StrictModeUnavailable`] if the message requests strict
/// KCP, [`EirnError::AuthenticationFailed`] if the KCP-Lite proof fails, or
/// [`EirnError::NoOneTimePrekeys`] if receiver state has no one-time prekey to
/// consume.
///
/// # Security
///
/// The receiver bundle must belong to the intended recipient and must not be
/// reused after its one-time prekey pool is depleted. A returned key is valid
/// only for the transcript represented by `msg0`.
pub fn receiver_handshake(bundle_b: &mut PrekeyBundle, msg0: &Msg0Prime) -> Result<[u8; 32]> {
    if msg0.kcp_mode != KcpMode::Lite {
        return Err(EirnError::StrictModeUnavailable);
    }

    let ctx = build_context(&msg0.pk_a, &bundle_b.identity_pk, &msg0.ek_a);
    if !kcp_lite_verify(msg0.pk_a.as_bytes(), &ctx, &msg0.kcp_lite_proof) {
        return Err(EirnError::AuthenticationFailed);
    }

    let (_, opk_sk) = bundle_b.consume_opk()?;
    let k1 = decaps(bundle_b.identity_sk(), &msg0.ct1);
    let k2 = decaps(bundle_b.signed_prekey_sk(), &msg0.ct2);
    let k3 = decaps(&opk_sk, &msg0.ct3);
    let k4 = naxos_decaps(bundle_b.signed_prekey_sk(), &msg0.ct4, &msg0.pk_a, &ctx);

    let secrets = [k1, k2, k3, k4];
    let ciphertexts = [
        msg0.ct1.as_bytes().as_slice(),
        msg0.ct2.as_bytes().as_slice(),
        msg0.ct3.as_bytes().as_slice(),
        msg0.ct4.as_bytes().as_slice(),
    ];
    Ok(derive_session_key(
        &secrets,
        msg0.pk_a.as_bytes(),
        bundle_b.identity_pk.as_bytes(),
        msg0.ek_a.as_bytes(),
        &ciphertexts,
        None,
    ))
}
